### ISSUE TITLE
Post idea : Vue.js Clean Components with the Composition API

### ISSUE THREAD

**Issue by ThomasFerro** [ThomasFerro/readmes#38]
> All about seperation of concerns

1. Make a component that manage too many things (data manipulation + templating);
2. Refactoring the component to extract data manipulation concerns from templating ones with renderless components;
3. Refactoring again to remove the unecessary component thanks to Vue 3 Composition API.

Dive deeper into the subject with [Vue Mastery's video on the topic presented by Evan You](https://www.vuemastery.com/courses/vue3-deep-dive-with-evan-you/logic-reuse/).

https://carbon.now.sh

**Comment by ThomasFerro** [comment]
# The state of Vue - Will the 3.0 still be approachable?

> Easy to get started with, but how hard to master?

It should not be long before we get our hands on Vue.js next major release! I cannot wait - and actually have not waited - to try the new features to build faster and cleaner applications thanks to the great work of Evan You and the community.

However, the pasts months had me wonder about the global reception of those new features, especially the Composition API. I sure am excited and ready to use it, but is everyone on the same page?

Let me try to explain why I am concerned about the future of the framework if it continues down that path.

## Approachable... at first

I remember more than four years ago, when I joined my first company and heard about Vue for the first time. Having tried only Angular.js in my years of study, I was anxious, not knowing if I could learn a completely new approach relatively fast.

Fast forward a few days later, I felt like I was up and running! The entry cost coming from Angular.js appeared quite cheep,. Everything seemed easier, more obvious.

When I say up and running, I did not mean that I knew every part of the framework and the ecosystem. I knew enough, however, to start contributing on the company's projects. The learning phase never stopped, as you may imagine. Every new feature to implement or bug to fix was an opportunity to teach myself good components and applications building practices.

## Digration about training

For now more than a year, I have been lucky enough to train people into Vue.js. The main target are developers already having a solid experience in Javascript so that we can focus on the framework. My goal is to provide them with the keys to build professional applications using this specific tool.

I was confident enough in my technical skills and knowledge, but not in my teaching skills. How to build a compelling and complete training? One that supports students up to a sufficient level of mastery?

With the help of a soft-skills coach and the support of my technical director, a first draft of the training was ready to be tested on the field.

Not long after, I was teaching Vue.js based on that first draft and I had a feeling that something was off. I gave the maximum amount of keys - knowledge about what the framework can do and how - and it seemed to connect with them. However, when asked to build a feature with no technical details, like in our day to day conversation with product owners and business experts, few where the ones who could provide a working solution.

What is it that makes Vue.js so hard to condensed in three days? *Condensed*... It may be it. Instead of *condensing* every feature into the training and risking to overwhelm people with non-essential information, why not *distill* those information through real world examples?

That is what I tried to apply in the latest version of the training, and it seems to work for most people. Of course, everyone has is own way of learning and there is no silver bullet, but I am convinced that it is a better way of teaching.

This training experience and the discovery of another framework, Svelte, had me wonder about my issues with Vue.js. It is certainly approachable at first, but what it is that confused me when learning Vue and seems to confuse those who want to get started?

## A hundred ways to achieve the same goal

Vue.js provides different ways to define *props* that your component accepts. For instance, here are three valid ways to tell that your component can receive an array of students:

```js
// 1. The simpliest solution, without any check

{
  props: [ 'students' ],
}

// 2. Type checked props

{
  props: {
    students: Array,
  },
};

// 3. Type checked props with fallback

{
  props: {
    students: {
      type: Array,
      // Another itch, the default value needs to be
      // inside a method for arrays and objects
      default() { return []; },
    },
  },
};
```

You have the same behaviour with the definition of *computed properties* and *watchers*.

This expandable API allows you to write less code if you do not need the extra options, which is nice... when you already know all the syntaxes.

When learning Vue, however, it tends to add a significant amount to the already big cognitive load.

As you may imagine, this is not going to get any better with the addition of the whole new Composition API. Newcomers will now be exposed with a whole new way or writing components logic **plus** the classical Options API.

I know that there is no silver bullet, that we have to choose between providing shorthands for developers or force everyone to use the expanded, more verbose version. I still think, however, that the following questions deserve to be asked:

Do we really *need* all of those options? Where do the convenience end and the pragmatism begin?

## Reactivity and references caveats in the hands on the developer

You may have seen that the Vue instance's `data` is often described as a function returning the actual data. Do you know why? According to [the documentation](https://vuejs.org/v2/guide/components.html#data-Must-Be-a-Function), you must do that so "[...]each instance can maintain an independent copy of the returned data object".

It sure sounds to me like a Javascript issue invading the framework's API, but I can get it. The thing that bugs me is that you *can* describe the data directly as an object to if you won't reuse the component.

Two syntaxes because of a references issue, bummer.

Going into the reactivity field, have you ever wonder why linters and compilers tend to get mad at you for not adding a `key` attribute when using the `v-for` directive? It is ["to maintain predictable behavior, such as object constancy in animations."](https://vuejs.org/v2/style-guide/#Keyed-v-for-essential) Again, I cannot help but think that this is not something we should manage as users.

I get that it could add complexity into the framework. Yet, I found it better to have technical complexities managed in the framework and leave the applications as clear as possible from them.

The same goes with the `deep` option in watchers, which allows you to not only watch the changes in the reference to an array or an object but also the changes in the array's elements or the object's properties. This last one, however, can be discussed since it can be useful to only listen to changes in the variable's reference.

Shouldn't those concerns be managed by the framework by default? Are we not using a framework to ignore those concerns in the first place? Should the developer know about implementation details in order to use a framework or library?

## Share your thoughts and opinion!

I know that I am just exposing points without bringing anything to the table. The goal is only to provide a context that we can use to build a constructive debate. I will dive deeper into my thoughts in the last article of this series.

What are your thoughts on the subject?

How did you first learn the basics of Vue.js?

Have you felt overwhelmed at first by the API's options?

Are you concerned about the addition of more APIs?

I will be glad to hear from you about it. On the next articles, we will actually discuss about Vue 3's Composition API and try to make the most out of it! 

--------

# Vue 3's Composition API and the segregation of concerns

> A novelty worth digging

I always think that it is best to put things in context. That is why I wrote [a first article](https://dev.to/thomasferro/the-state-of-vue-will-the-3-0-still-be-approachable-4mgc) about my take on the state of Vue before the 3.0, which is currently available as a release candidate.

However, the main subject of this series is one of Vue 3's new feature: the Composition API. It is the one that I anticipate the most and it is finally time we discuss it!

## Shiny new toys

This article should be the shortest of the series, since the subject has already been discussed [many](https://auth0.com/blog/getting-started-with-vue-3-composition-api/) [times](https://composition-api.vuejs.org/) by people way more interesting and smart than I am.

The Composition API was created to answer two common issues that occur when your Vue application starts to get big.

### Code organization

Have you ever had to maintain really big components, with complex logic implying a lot of `data`, `computed`, `methods`, etc? When trying to read those types of components, the main issue is to keep track of what everything is doing and how do they interact with each other. With the current Options API, you have to navigate back and forth inside the Vue instance, which causes a great cognitive load.

Vue 3 tries to resolve this issue by adding a new method to the Vue instance, the [`setup`](https://composition-api.vuejs.org/api.html#setup). This method can be seen as **the component's entry point**, being called before the `beforeCreated` hook and receiving the `props` as an argument. The **returned value is an object containing every available information for the template to use**.

This is the place where you will write all of your component logic, no matter if we are talking about `data`, `computed`, `watcher`, etc.

There is still a missing piece to this puzzle, how do we write `data`, `computed`, `watcher`, `methods` and more in that new `setup` method?

Vue 3 **provides a new tool to create those reactive data and more: the Reactivity API**.

It may not be relevant for now, but here is a small example of how to create a reactive data using the Reactivity API:

```js
import { ref } from 'vue';

const count = ref(0);

// Accessing ref's value in JS
console.log('Count:', count.value)

// Modifying the value
count.value += 1
```

As you can see, you have to explicitly access the `ref`'s value when manipulating it in JS. It bugs me a little, but you won't have to do so in the template and can directly access the value, as we will see later.

Please take a look at the [Reactivity API's reference](https://composition-api.vuejs.org/api.html#reactivity-apis) for more information about what is contained in it.

Ok, but how do those two APIs fit together? Let us see that with an example. First, we will write it using the Options API, then we will do it *à la Vue 3*.

Let us say that we have a component managing the loading and display of blog posts, a minimalist version of it could be like this:

```js
export default {
    name: 'blog-posts',
    data() {
        return {
            posts: [],
            loadingStatus: '',
            error: '',
        };
    },
    computed: {
        blogPostsLoading() {
            return this.loadingStatus === 'loading';
        },
        blogPostsLoadingError() {
            return this.loadingStatus === 'error';
        },
    },
    methods: {
        loadBlogPosts() {
            this.loadingStatus = 'loading';
            fetch(process.env.VUE_APP_POSTS_URL)
                .then((response) => {
                    if (!response.ok) {
                        throw new Error(response.status);
                    }
                    return reponse.json();
                })
                .then((posts) => {
                    this.posts = posts;
                    this.loadingStatus = 'loaded';
                })
                .catch((error) => {
                    this.error = error;
                    this.loadingStatus = 'error';
                });
        },
    },
    created() {
        this.loadBlogPosts();
    },
}
```

Using the new provided tools, we can put all of the logic in the `setup`:

```js
import { ref, computed } from 'vue';

export default {
    name: 'blog-posts',
    setup() {
        const loadingStatus = ref('');
        const error = ref('');
        const posts = ref([]);

        const blogPostsLoading = computed(() => {
            return loadingStatus.value === 'loading';
        });
        const blogPostsLoadingError = computed(() => {
            return loadingStatus.value === 'error';
        });

        const loadBlogPosts = () => {
            loadingStatus.value = 'loading';
            fetch(process.env.VUE_APP_POSTS_URL)
                .then((response) => {
                    if (!response.ok) {
                        throw new Error(response.status);
                    }
                    return reponse.json();
                })
                .then((fetchedPosts) => {
                    posts.value = fetchedPosts;
                    loadingStatus.value = 'loaded';
                })
                .catch((apiError) => {
                    error.value = apiError;
                    loadingStatus.value = 'error';
                });
        };

        // Return every information to be use by the template
        return {
            loadingStatus,
            // You can rename those information if needed
            loadingError: error,
            loadBlogPosts,
            blogPostsLoading,
            blogPostsLoadingError,
        };
    },
};
```

It can seem not so useful in a component with few logic, but it already helps developers keep track of the different pieces without scrolling between the Vue instance's options. We will see later in this article and the next ones how to leverage the most out of it.

We can also extract the logic by creating an ES module (`posts.js`) managing the data and exposing the useful information:

```js
import { ref, computed } from 'vue';

export const useBlogPosts = () => {
    const loadingStatus = ref('');
    const error = ref('');
    const posts = ref([]);

    const blogPostsLoading = computed(() => {
        return loadingStatus.value === 'loading';
    });
    const blogPostsLoadingError = computed(() => {
        return loadingStatus.value === 'error';
    });

    const loadBlogPosts = () => {
        loadingStatus.value = 'loading';
        fetch(process.env.VUE_APP_POSTS_URL)
            .then((response) => {
                if (!response.ok) {
                    throw new Error(response.status);
                }
                return reponse.json();
            })
            .then((fetchedPosts) => {
                posts.value = fetchedPosts;
                loadingStatus.value = 'loaded';
            })
            .catch((apiError) => {
                error.value = apiError;
                loadingStatus.value = 'error';
            });
    };
    
    // Return every information to be use by the consumer (here, the template) 
    return {
        loadingStatus,
        // You can rename those information if needed
        loadingError: error,
        loadBlogPosts,
        blogPostsLoading,
        blogPostsLoadingError,
    };
}
```

Our component will then only manage the template based on data provided by the module. Complete separation of concerns:

```js
import { useBlogPosts } from './posts.js';

export default {
    name: 'blog-posts',
    setup() {
        const blogPostsInformation = useBlogPosts();
        return {
            loadingStatus: blogPostsInformation.loadingStatus,
            loadingError: blogPostsInformation.loadingError,
            loadBlogPosts: blogPostsInformation.loadBlogPosts,
            blogPostsLoading: blogPostsInformation.blogPostsLoading,
            blogPostsLoadingError: blogPostsInformation.blogPostsLoadingError,
        };
    },
};
```

Again, it helps clarify your code and decouple the intent from the implementation, which is always nice.

You may already have thought about it, but this way of creating modules can help us reuse logic!

### Logic reuse

We already have some tools that help us create logic to be used by many components. [Mixins](https://vuejs.org/v2/guide/mixins.html), for instance, let write you any Vue instance's options to be injected in one or many components.

These approaches share one downside, they lack clarity.

You never clearly know which mixin imported which option unless you read through all of them. It can easily become a nightmare for developers trying to understand how the components work, having to navigate through the Vue instance and the globally and locally injected mixins. Moreover, mixins' options can collide with each others, leading to a knot bag, not to say a mess.

With the Composition API, any component can pick what he needs from different modules. Clearly stated in the `setup` method, the developers can see what was taken from where and even rename variable if it helps understand the intent better.

I think that clarity is a, if not *the*, most important concern when writing applications to be actively maintain for years. The Composition API gives us the tool to do so in an elegant and practical way.

### Wait, there is more?

The two main goals look pretty achieved to me, but the Composition API should not be reduced to those two concerns.

It will also **benefit the testability of our applications**, let me explain how.

Before Vue 3 we had to instantiate the component, inject mocked dependencies if needed, and only then start hacking around assertions. This way of testing **can lead to tests suites strongly coupled to the actual implementation**. The kind of tests that age poorly and can do way more harm than good. 

Now we can create ES modules encapsulating the domain logic and exporting the data and methods to be used. Those modules will be written in almost-pure Javascript since they will still use Vue's API, but not in the context of a component.

Our tests can simply consume the exported information, just like our components will!

## The art of teaching

You may have noticed by the fact that I am writing a whole series about it, I am really excited about this new API. It scratches itches that I've had for a long time, trying to apply clean code principles into my front-end applications. I think it will help us tremendously increase the quality of our components and applications if well used.

However, the Composition API is an advanced concept. I do not think that it should replace the actual way of writing components. Moreover, we will still encounter legacy code written before Vue 3, so our previous knowledge is still useful.

I already discuss this issue in the previous article, but it is really important to keep this in mind: not everyone is lucky enough to discover the 3.0 after two years of practicing Vue almost daily.

Some people will start using Vue with the 3.0, and a whole new API like this one adds a lot to the already big entry cost. A newcomer now have to understand it in addition to the "classical" Options API.

**How do you think that the new API should be introduced to newcomers?** I personally think that it should be just like Vuex or Vue Router, introduced later as an advanced tool. It should be added on top of a solid foundations of knowledge and used practically.

## Once again, share your thoughts!

What do you think about the new Composition API?

Are you ready to use it as soon as the 3.0 is released?

Please let everyone know and let us discuss all that :)

Now that the subject is theoretically introduced, what is next? I will get my hands dirty and try the make the most out of the Composition API, starting in the next article with a refactoring session!

**Comment by ThomasFerro** [comment]
# Composition API v Renderless Components - Let's use Vue 3's features to clean our components!

> Iteratively and incrementally improve components quality

Making the perfect component in the first draft is impossible. Impossible because you do not know exactly what will be needed before actually building the component. Impossible also because you will always learn new, more effective ways of doing things.

Too many times have I overengineered, trying to make components that no one would ever need to modify or fix, components that was meant to meet all present and future use cases.

Let me tell you the best place that I found for those components: in a post-mortem.

*Why is he telling me all that*, you may ask yourself. I wanted to introduce this article with this little digression to discuss the importance of iterations.

First, build the minimum viable product, the basic implementation that works and does what is expected. Only then starts the refinement phase to ship a clean and maintainable code.

You do not know if you are building the right thing until you try it. Ship it fast and get feedback.

It is a concept you should be familiar with [when doing TDD](https://medium.com/@t.ferro184/tdd-in-action-debouncer-in-go-b0f7d7d75931) or even if you recognize yourself in the Agile values.

This article follows the same pattern, we will start with a component that works, even though it is far from being maintainable. Then we will incrementally improve it, without the new Composition API in the first place so we will be able to compare with previously existing tools.

I will not discuss the mostly important matter of tests in this article. The reason being that I am not confident enough on the subject of Front-End testing to give you my opinion. I might dig into the subject in the future, but for now I leave you with a few resources:

- [Vue Test Utils - the official unit testing utility library for Vue.js](https://vue-test-utils.vuejs.org/);
- [Testing Library - a collection of utilities that encourage "good" testing practices](https://testing-library.com/). I haven't tried it yet, but the promise is good enough for me to share it with you.

## The legacy component

Before starting any refactoring, we need to understand what we are working with.

We will be creating a TODO-list with only a few features:

- Listing the tasks to do;
- Creating a new task;
- Tag a task as finishing.

The first thing we want to do is to make the application work, so let us do it!

```html
<template>
    <h1>My TODO list! ({{ doneCount }} / {{ totalCount }})</h1>

    <!-- Error management -->
    <p v-if="loadingError">
        {{ loadingError }}
        <button @click="loadTodos">Reload</button>
    </p>

    <ul v-else>
        <li v-for="todo in todoList" :key="todo.id">
            {{ todo.content }}
            <button @click="achieveATodo(todo.id)">Validate</button>
        </li>
    </ul>
    <form @submit.prevent="() => addTodo(newTodoContent)">
        <label>
            What do you have to do?
            <input v-model="newTodoContent">
        </label>
        <input type="submit" value="Create">
        <!-- Error management -->
        <p v-if="todoCreationError">{{ todoCreationError }}</p>
    </form>
</template>

<script>
export default {
    name: 'todo-list',
    data() {
        return {
            loadingError: '',
            todoList: [ ],
            newTodoContent: '',
            todoCreationError: '',
        };
    },
    computed: {
        doneCount() {
            return this.todoList.filter(todo => todo.done).length;
        },
        totalCount() {
            return this.todoList.length;
        },
    },
    methods: {
        loadTodos() {
            this.loadingError = '';
            fetch(import.meta.env.VITE_TODOS_URL)
                .then((response) => {
                    if (!response.ok) {
                        throw new Error('An error has occurred while loading todos');
                    }
                    return response.json();
                })
                .then((todos) => {
                    this.todoList = todos;
                })
                .catch((error) => {
                    this.loadingError = error;
                });
        },
        achieveATodo(id) {
            // [...] Call the API to achieve the task
        },
        addTodo(content) {
            this.todoCreationError = '';
            fetch(import.meta.env.VITE_TODOS_URL, {
                method: 'post',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ content })
            })
                .then((response) => {
                    if (!response.ok) {
                        throw new Error('An error has occurred while creating todo');
                    }
                    return response.json();
                })
                .then((newTodo) => {
                    this.todoList = [
                        ...this.todoList,
                        newTodo,
                    ]
                })
                .catch((error) => {
                    this.todoCreationError = error;
                });
        }
    },
    created() {
        this.loadTodos();
    },
}
</script>
```

Great, we have a working component. The client is happy since he can try the application even though it is a *work in progress*. Now that we know that his expectation are met, it is time to meet ours.

Listing the pros and cons of the component can be useful in this case. It will allow us to have a complete picture of the state of the component and we will be able to prioritize the tasks to be done.

On the bright side, the component works. He has every needed features and we managed to build it relatively fast. On the other hand, we have a lot to improve before shipping it:

- It has too many responsibilities (data fetching, business rules applying to the data, actions to add and modify data, the display of all of those information);
- Nothing here is reusable;
- It seems hard to maintain, at least it is hard to read through (a hundred lines, without the task achievement logic).

Let us iterate a few times until we are happy about the result!

## Before Vue 3, a first step toward clean components

This is kind of a bonus section where we will refactor the component using present tools. You can skip it if you are only interested in the Composition API or if you are not familiar with the advanced concept of [*scoped slots*](https://vuejs.org/v2/guide/components-slots.html#Scoped-Slots). On the other hand, I do think that it is an interesting pattern to see if not to learn and use.

I had the chance to work with a friend and former colleague, [Edouard](https://twitter.com/ecattez) [Cattez](https://github.com/ecattez), on a project for a big French retailer. The Front-End of this project was made, you guessed it, using Vue.js.

We had an issue with the code base that we could not name. For several months we worked hard on it, but could not figure what it was that make so difficult to add or modify features.

This was about the same time where I started to really dive into the concepts of Clean Code, Clean Architecture and Software Craftsmanship.

One day, when talking to that friend, we finally were able to find the underlying issue, our code base lacked separation of concerns.

Every components in our application started to get quite big since they managed their template, data managing and styles. This way of writing components can work fine, as long as it does not get out of hands.

Our components, however, managed a lot of business logic and associated templates. It causes a great amount of cognitive load to read through since the components held the intent *and* the implementation of the business logic.

We needed a way to separate the concerns, to have the business logic in one place and the templates in another. We could drop the *Single File Components* or even write mixins, but those solutions sounded wrong in our context.

The issue was not that the template and the data managing were in the same file. It had more to do with the fact that we mixed the intent and the implementation. Like an application without *interfaces*, only implementations.

This is were we found out about the great article and pattern from [*Adam Wathan*](https://adamwathan.me), [**Renderless Components in Vue.js**](https://adamwathan.me/renderless-components-in-vuejs/).

I will not dig too deep inside the matter since his article already explain it all. Just know that it works by creating a **renderless component**, responsible for the data managing. This renderless component then provides information for the "*view component*" to use thanks to **scoped slots**.

How could we apply this pattern in our TODO-list? Let us first try to extract the logic inside a renderless component, named **`TodoListManager`**:


```html
<!-- No template tag, we will use a render function -->
<script>
export default {
    name: 'todo-list-manager',
    data() {
        return {
            loadingError: '',
            todoList: [ ],
            todoCreationError: '',
        };
    },
    computed: {
        doneCount() {
            return this.todoList.filter(todo => todo.done).length;
        },
        totalCount() {
            return this.todoList.length;
        },
    },
    methods: {
        loadTodos() {
            this.loadingError = '';
            fetch(import.meta.env.VITE_TODOS_URL)
                .then((response) => {
                    if (!response.ok) {
                        throw new Error('An error has occurred while loading todos');
                    }
                    return response.json();
                })
                .then((todos) => {
                    this.todoList = todos;
                })
                .catch((error) => {
                    this.loadingError = error;
                });
        },
        achieveATodo(id) {
            // [...] Call the API to achieve the task
        },
        addTodo(content) {
            this.todoCreationError = '';
            fetch(import.meta.env.VITE_TODOS_URL, {
                method: 'post',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ content })
            })
                .then((response) => {
                    if (!response.ok) {
                        throw new Error('An error has occurred while creating todo');
                    }
                    return response.json();
                })
                .then((newTodo) => {
                    this.todoList = [
                        ...this.todoList,
                        newTodo,
                    ]
                })
                .catch((error) => {
                    this.todoCreationError = error;
                });
        }
    },
    created() {
        this.loadTodos();
    },
    render() {
        // Only display the content inside of the default slot, with every needed information
        return this.$slots && this.$slots.default && this.$slots.default({
            loadTodos: this.loadTodos,
            loadingError: this.loadingError,
            todoList: this.todoList,
            doneCount: this.doneCount,
            totalCount: this.totalCount,
            achieveATodo: this.achieveATodo,
            addTodo: this.addTodo,
            todoCreationError: this.todoCreationError,
        });
    },
}
</script>
```

While the view component could be like this one:

```html
<template>
    <!-- Use our renderless component -->
    <!-- You can see that not only data are provided but also methods, computed, etc -->
    <todo-list-manager v-slot="{
        loadTodos,
        loadingError,
        todoList,
        doneCount,
        totalCount,
        achieveATodo,
        addTodo,
        todoCreationError,
    }">
        <!-- Here, we can use every reactive information provided by the renderless component -->
        <h1>My TODO list! ({{ doneCount }} / {{ totalCount }})</h1>

        <!-- Error management -->
        <p v-if="loadingError">
            {{ loadingError }}
            <button @click="loadTodos">Reload</button>
        </p>

        <ul v-else>
            <li v-for="todo in todoList" :key="todo.id">
                {{ todo.content }}
                <button @click="achieveATodo(todo.id)">Validate</button>
            </li>
        </ul>
        <form @submit.prevent="() => addTodo(newTodoContent)">
            <label>
                What do you have to do?
                <!-- newTodoContent may come from the view component or the renderless one -->
                <input v-model="newTodoContent">
            </label>
            <input type="submit" value="Create">
            <!-- Error management -->
            <p v-if="todoCreationError">{{ todoCreationError }}</p>
        </form>
    </todo-list-manager>
</template>

<script>
// [...]
</script>
```

We could go even further by extracting the API call inside a JS module, creating a generic loading and error display management component, etc. Those enhancements are out of the scope of the article, but still great to do. What we can do now, however, is to keep iterating on the renderless component.

Our `TodoListManager` seems greatly filled to me. What if we only need to list the tasks? What if we only need to create a new one?

We could ignore the data exposed by the renderless component that we do not need. However, I find it more clear to explicitly use the renderless component responsible for the creation of a task and / or the one responsible for the listing. Here is how we can achieve that.

First, the creation logic is extracted in a new renderless component, **`TodoCreationManager`**:

```html
<script>
export default {
    name: 'todo-creation-manager',
    data() {
        return {
            todoCreationError: '',
        };
    },
    emits: [ 'todo-created' ],
    methods: {
        addTodo(content) {
            this.todoCreationError = '';
            fetch(import.meta.env.VITE_TODOS_URL, {
                method: 'post',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({ content })
            })
                .then((response) => {
                    if (!response.ok) {
                        throw new Error('An error has occurred while creating todo');
                    }
                    return response.json();
                })
                .then((newTodo) => {
                    // We don't have any reference to the list here
                    // We can, however, send an event with the created task
                    this.$emit('todo-created', newTodo)
                })
                .catch((error) => {
                    this.todoCreationError = error;
                });
        }
    },
    render() {
        return this.$slots && this.$slots.default && this.$slots.default({
            addTodo: this.addTodo,
            todoCreationError: this.todoCreationError,
        });
    },
}
</script>
```

Our **`TodoListManager`** component is now only responsible for the fetching of the tasks list.

Then, in our view component, we need to nest the two renderless component and use the logic from them both in the template:

```html
<template>
    <!-- Use our renderless component -->
    <!-- You can see that not only data are provided but also methods, computed, etc -->
    <todo-list-manager v-slot="{
        loadTodos,
        loadingError,
        todoList,
        doneCount,
        totalCount,
        achieveATodo,
        todoCreated,
    }">
        <!-- A second renderless component, managing the creation of a task -->
        <!-- 
            When this component notify us that a new task is created,
            we can add it directly to the list by calling a method
            on the todo-list-manager renderless component
         -->
        <todo-creation-manager
            v-slot="{
                addTodo,
                todoCreationError,
            }"
            @todo-created="todoCreated"
        >
            <!-- Here, we can use every reactive information provided by the renderless component -->
            <h1>My TODO list! ({{ doneCount }} / {{ totalCount }})</h1>

            <!-- Error management -->
            <p v-if="loadingError">
                {{ loadingError }}
                <button @click="loadTodos">Reload</button>
            </p>

            <ul v-else>
                <li v-for="todo in todoList" :key="todo.id">
                    {{ todo.content }}
                    <button @click="achieveATodo(todo.id)">Validate</button>
                </li>
            </ul>
            <form @submit.prevent="() => addTodo(newTodoContent)">
                <label>
                    What do you have to do?
                    <!-- newTodoContent may come from the view component or the renderless one -->
                    <input v-model="newTodoContent">
                </label>
                <input type="submit" value="Create">
                <!-- Error management -->
                <p v-if="todoCreationError">{{ todoCreationError }}</p>
            </form>
        </todo-creation-manager>
    </todo-list-manager>
</template>
```

-----

It is a pattern that I adopted for every component with complex business logic. It helps keeping your view component clean and concise. However, since it is based on a renderless *component*, it adds one to the components tree each time you use it. **It is also worth noting that it is an advance pattern that adds to the entry cost of your code base**.

How is this elegant solution compared to the new Composition API? Let us find out. 

## Refactoring in Vue 3 with the Composition API

In this section, I will assume that you are already familiar with the intent and the basic syntax of the Composition API.

I made [an article introducing the API in case you never heard about it](https://dev.to/thomasferro/vue-3-s-composition-api-and-the-segregation-of-concerns-1k5c). Please read it first if you are afraid of being confused by the syntax.

We have two implemented features:

- Fetch the todo list;
- Add a new one.

You can try to follow the same pattern while implementing the task achievement if you want.

Let us start with the list fetching. First, we will create a new ES module with a method that contains every pieces of information about the todo list. It is basically the same as the data inside the carryall component, but with a different syntax:

```js
import { ref, computed } from 'vue';

export const useTodoList = () => {
    // First, we create the reactive data and computed
    const todoList = ref([ ]);
    const doneCount = computed(() => {
        return todoList.value.filter(todo => todo.done).length;
    });
    const totalCount = computed(() => {
        return todoList.value.length;
    });

    const loadingError = ref('');

    // Then we create the method that will manipulate those data
    const loadTodos = () => {
        loadingError.value = '';
        fetch(import.meta.env.VITE_TODOS_URL)
            .then((response) => {
                if (!response.ok) {
                    throw new Error('An error has occurred while loading todos');
                }
                return response.json();
            })
            .then((todos) => {
                todoList.value = todos;
            })
            .catch((error) => {
                loadingError.value = error;
            });
    }

    const achieveATodo = (id) => {
        // [...] Call the API to achieve the task
        // Move it in a new method useTodoAchiever
    };

    // This method will be useful soon
    const todoCreated = (newTodo) => {
        todoList.value = [
            ...todoList.value,
            newTodo
        ]
    }
    
    // Finaly, we return the information that could be useful for our clients
    return {
        todoList,
        doneCount,
        totalCount,
        loadingError,
        loadTodos,
        achieveATodo,
        todoCreated,
    }
}
```

These information will be consumed by our view component's `setup` method. Here is the **`TodoList`**:

```html
<template>
    <!-- The template remains untouched -->
</template>

<script>
import { useTodoList } from './index.js';

export default {
    name: 'todo-list',
    setup() {
        // You cannot destructure the returned value here or you will loose Vue's reactivity
        const todoListData = useTodoList();

        todoListData.loadTodos();

        return {
            todoList: todoListData.todoList,
            doneCount: todoListData.doneCount,
            totalCount: todoListData.totalCount,
            loadingError: todoListData.loadingError,
            loadTodos: todoListData.loadTodos,
            achieveATodo: todoListData.achieveATodo,
        }
    },
}
</script>
```

We can now do the same with the task creation process:

```js
export const useTodoCreation = ({
    // Method called when a todo is created
    onTodoCreated = () => {},
}) => {
    // Create the reactive data
    const todoCreationError = ref('');

    // The method used to create a new task
    const addTodo = (content) => {
        todoCreationError.value = '';
        fetch(import.meta.env.VITE_TODOS_URL, {
            method: 'post',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ content })
        })
            .then((response) => {
                if (!response.ok) {
                    throw new Error('An error has occurred while creating todo');
                }
                return response.json();
            })
            .then(onTodoCreated)
            .catch((error) => {
                todoCreationError.value = error;
            });
    }

    // Return the needed information
    return {
        todoCreationError,
        addTodo,
    }
}
```

Plug what is needed inside the view component:

```html
<script>
import { ref } from 'vue';
import { useTodoList, useTodoCreation } from './index.js';

export default {
    name: 'todo-list',
    setup() {
        // You cannot destructure the returned value here or you will loose Vue's reactivity
        const todoListData = useTodoList();
        const todoCreationData = useTodoCreation({
            // Plug the method that will update the list when a task is created
            onTodoCreated: todoListData.todoCreated,
        });
        const newTodoContent = ref('');

        todoListData.loadTodos();

        return {
            todoList: todoListData.todoList,
            doneCount: todoListData.doneCount,
            totalCount: todoListData.totalCount,
            loadingError: todoListData.loadingError,
            loadTodos: todoListData.loadTodos,
            achieveATodo: todoListData.achieveATodo,
            todoCreationError: todoCreationData.todoCreationError,
            addTodo: todoCreationData.addTodo,
            newTodoContent,
        }
    },
}
</script>
```

The last thing we can do is to create reusable component for the display of a task and for the creation form.

```html
<!-- TodoCreation.vue -->
<template>
    <form @submit.prevent="() => addTodo(newTodoContent)">
        <label>
            What do you have to do?
            <input v-model="newTodoContent">
        </label>
        <input type="submit" value="Create">
        <!-- Error management -->
        <p v-if="creationError">{{ creationError }}</p>
    </form>
</template>

<script>
export default {
    name: 'todo-creation',
    // Declare what events will our component emit
    emits: [
        'create-todo',
    ],
    props: {
        creationError: String,
    },
    data() {
        return {
            newTodoContent: '',
        }
    },
    methods: {
        addTodo(content) {
            this.$emit('create-todo', { content });
        }
    },
}
</script>
```

```html
<!-- TodoDisplay.vue -->
<template>
    {{ content }}
    <button @click="achieveTodo()">Validate</button>
</template>

<script>
export default {
    name: 'todo-display',
    emits: [
        'achieve-todo',
    ],
    props: {
        content: String,
    },
    methods: {
        achieveTodo() {
            this.$emit('achieve-todo');
        }
    },
}
</script>
```

```html
<!-- TodoList.vue -->
<template>
    <!-- Here, we can use every reactive information provided by the renderless component -->
    <h1>My TODO list! ({{ doneCount }} / {{ totalCount }})</h1>

    <!-- Error management -->
    <p v-if="loadingError">
        {{ loadingError }}
        <button @click="loadTodos">Reload</button>
    </p>

    <ul v-else>
        <li v-for="todo in todoList" :key="todo.id">
            <todo-display
                :content="todo.content"
                @achieve-todo="() => achieveATodo(todo.id)"
            ></todo-display>
        </li>
    </ul>
    <todo-creation
        :creation-error="todoCreationError"
        @create-todo="addTodo"
    ></todo-creation>
</template>

<script>
    // [...]
</script>
```

-----

This is even cleaner than the solution using *renderless component* to me since it does not add components to the tree. The Composition API allows for a strict segregation of concerns. Our components can use business logic without knowing implementation details.

It will, however and just like the *renderless components*, adds to the entry cost of our projects. That is why I will use it pragmatically and try to make it the most readable possible for newcomers. For instance, in this application, I only used it in the **`TodoList`** component.

## On the next episode...

I hope that this series helps you better understand the benefits of Vue 3's Composition API! Please send feedback of your journey learning to use this new toy :)

What do you think about those two methods? Which one is the clearer, the one that you will start to use?

Next on the series, I will show you how to get rid of Vuex thanks to the Composition API.

**Comment by ThomasFerro** [comment]
# Emancipate yourself from Vuex with Vue 3's Composition API

> Is there still room for Vuex with Vue 3?

Before wrapping up this series about *Vue 3* and the *Composition API*, I wanted to show you one last use-case that I found interesting. If you did not already, please take a look at [my introduction of the Composition API](https://dev.to/thomasferro/vue-3-s-composition-api-and-the-segregation-of-concerns-1k5c) so that you won't get lost with the syntax.

This article is especially for the people who have already learned *Vuex* and tend to use it in every place where they manage data.

If you do not fit into this category or do not even know what *Vuex* is, here is a concise introduction.

## One store to rule them all

According to the documentation, *Vuex* is "a state management pattern + library for Vue.js applications". Think of it as the place to store and manipulate reactive data outside of a component before we had the Reactivity and the Composition APIs.

I cannot recommend you enough to watch **Vue Mastery**'s introduction on the subject, available on [the library's home page](https://vuex.vuejs.org/).

To sum up greatly, you can use this library to externalize reactive data shared by components far from each other in the components tree.

Instead of communicating by sending props down the tree and emitting events up, you can use a *Vuex store*. This way, your components are always up to date when one of them modify the store's state.

These two schemas are from [*Vue Mastery*'s introduction](https://www.vuemastery.com/courses/mastering-vuex/intro-to-vuex/). First you can see a complex component tree with a lot of events and props to manage in order to make the two leaf components communicate:

![Complex component tree communication - Vue Mastery](https://firebasestorage.googleapis.com/v0/b/vue-mastery.appspot.com/o/flamelink%2Fmedia%2F1578371882429_1.png?alt=media&token=2c411c9f-d5cf-4404-8009-00b73e24a622)

Here we use a Vuex store to simplify this communication:

![Components communication with Vuex - Vue Mastery](https://firebasestorage.googleapis.com/v0/b/vue-mastery.appspot.com/o/flamelink%2Fmedia%2F1578371889694_2.png?alt=media&token=f9cc01f0-4644-4867-92ad-17ccd6cc7a6e)

A *Vuex store* is composed of these different parts:

- [`state`](https://vuex.vuejs.org/guide/state.html): The place where your reactive data lives. **You cannot directly modify the state**. To do so, you have to *commit* **mutations**;
- [`getters`](https://vuex.vuejs.org/guide/getters.html): Just like computed properties, this is a way to expose the state in a different shape. They are usually used to avoid rewriting logic inside every component consuming the store;
- [`mutations`](https://vuex.vuejs.org/guide/mutations.html): The only way to modify the state is by **committing mutations**. They should be synchronous and the smallest possible;
- [`actions`](https://vuex.vuejs.org/guide/actions.html): For asynchronous treatments or for logic that implies modifying many elements in the state, we can **dispatch actions**;
- [`modules`](https://vuex.vuejs.org/guide/modules.html): In order to split the state, we can create stand-alone stores, the modules.

The state is represented as a stack of mutations that you can replay or just analyze in-depth:

![Mutations stack shown in Vue Devtools](todo)

This is just a theoretical introduction and it is not enough to get started. Please read through [the documentation for more information](https://vuex.vuejs.org/).

### My issue with Vuex

*Vuex* is just like any technology, it comes with a cost. First, the price of learning the library. It usually takes me half a day to only introduce the subject in the trainings that I gave. You can add a couple of days of practicing before actually tame the beast.

Secondly, when using *Vuex*, you tend to lost the concept of **data responsibility**. No component is responsible for the data when everyone can modify the store's state. This way of thinking usually leads to application that are hard to maintain and debug. It is hard to keep track of whom did the mutations and why, even when using great tools such as [Vue Devtools](https://github.com/vuejs/vue-devtools).

When I say "usually", I generalize on purpose. They are applications where *Vuex* is used pragmatically and where the code base is still easily maintainable and scalable. However, I tended to overuse *Vuex* when I first learned it and I think that I am not the only one.

My take is to almost never use *Vuex*, especially when a simple "Parent-Child" communication pattern is enough. You will save yourself from long hours of debugging and headaches.

Where to use it, then? There are use-cases where it comes handy. Or should I say "where it used to came handy", now that we have the Composition API. Let me explain with an example, a sample application with user information displayed and editable.

## User data management using Vuex

I will not be covering Vuex installation in this article, please follow [the documentation](https://vuex.vuejs.org/installation.html) if you need to.

Let us first take a look at what we will be building. We are not creating an entire Web application, it is way out of this article's scope. We will, however, build common pieces that you will certainly encounter if you have not already. Those two pieces are the following:

1. A `Header` component, displaying the user nickname and his profile picture;
2. A `UserSettings` component where the data will be editable.

Here, using *Vuex* is overkill. Just imagine that the rest of the application is ignored, that we have a complex component tree and *Vue Router* installed.

The actual API call will be externalize in the `api.js` file. Just know that it returns a `Promise`, like `fetch` or `axios` would have.

Let us start our implementation with the store's user module:

```js
import { loadUserInfo, saveNewUserInfo } from './api';

const AVAILABLE_STATUS = {
    LOADING: 'LOADING',
    UPDATING: 'UPDATING',
    ERROR: 'ERROR',
};

export const user = {
    namespaced: true,
    state() {
        return {
            nickname: '',
            pictureUrl: '',
            status: '',
        };
    },
    getters: {
        isLoading: state => state.status === AVAILABLE_STATUS.LOADING,
        isUpdating: state => state.status === AVAILABLE_STATUS.UPDATING,
        errorOccurred: state => state.status === AVAILABLE_STATUS.ERROR,
    },
    mutations: {
        changeStatus(state, newStatus) {
            state.status = newStatus;
        },
        changeNickname(state, newNickname) {
            state.nickname = newNickname;
        },
        changePicture(state, newPicture) {
            state.pictureUrl = newPicture;
        },
    },
    actions: {
        // Called by the "App" component to ensure that the initial data are loaded
        load({ commit }) {
            commit('changeStatus', AVAILABLE_STATUS.LOADING);
            loadUserInfo()
                .then(({ nickname, pictureUrl }) => {
                    commit('changeNickname', nickname)
                    commit('changePicture', pictureUrl)
                    commit('changeStatus', '');
                })
                .catch(() => {
                    commit('changeStatus', AVAILABLE_STATUS.ERROR);
                })
        },
        update({ commit }, newUser) {
            commit('changeStatus', AVAILABLE_STATUS.UPDATING);
            saveNewUserInfo(newUser)
                .then(({ nickname, pictureUrl }) => {
                    commit('changeNickname', nickname)
                    commit('changePicture', pictureUrl)
                    commit('changeStatus', '');
                })
                .catch(() => {
                    commit('changeStatus', AVAILABLE_STATUS.ERROR);
                })
        },
    },
};
```

Here we have two important things. First, the state to consume with the nickname and the picture's url. We also have the possibility to modify the profile thanks to the `update` *action*.

A loading status is also managed in the store, which allows components to display the appropriate message to the user.

The header component can now consume the store's data:

```html
<template>
    <header>
        <template v-if="isLoading">
            User information are loading
        </template>
        <template v-else-if="isUpdating">
            User information are updating
        </template>
        <template v-else-if="errorOccurred">
            Unable to manage user information
        </template>
        <template v-else>
            Welcome {{ nickname }}
            <img :src="pictureUrl" alt="User picture" class="user-picture">
        </template>
    </header>
</template>

<script>
import { mapState, mapGetters } from 'vuex';

export default {
    name: 'app-header',
    computed: {
        ...mapState({
            nickname: state => state.user.nickname,
            pictureUrl: state => state.user.pictureUrl,
        }),
        ...mapGetters({
            isLoading: 'user/isLoading',
            isUpdating: 'user/isUpdating',
            errorOccurred: 'user/errorOccurred',
        }),
    },
}
</script>

<style >
.user-picture {
    height: 40px;
    width: 40px;
    border-radius: 50%;
}
</style>
```

Last, the `UserSettings` component will do just the same and will use the action when the user validate his modifications:

```html
<template>
    <form @submit.prevent="updateUser">
        <label>
            Nickname
            <input type="text" v-model="newNickname">
        </label>
        <label>
            Picture url
            <input type="text" v-model="newPicture">
        </label>
        <input type="submit" value="Validate changes" :disabled="formDisabled">
        <p v-if="errorOccurred">An error has occurred while managing user information...</p>
    </form>
</template>

<script>
import { mapState, mapGetters } from 'vuex';

export default {
    name: 'user-settings',
    data() {
        return {
            newNickname: '',
            newPicture: '',
        };
    },
    computed: {
        ...mapState({
            nickname: state => state.user.nickname,
            pictureUrl: state => state.user.pictureUrl,
        }),
        ...mapGetters({
            isLoading: 'user/isLoading',
            isUpdating: 'user/isUpdating',
            errorOccurred: 'user/errorOccurred',
        }),
        formDisabled() {
            return this.isLoading || this.isUpdating
        },
    },
    watch: {
        nickname: {
            handler() {
                this.newNickname = this.nickname;
            },
            immediate: true,
        },
        pictureUrl: {
            handler() {
                this.newPicture = this.pictureUrl;
            },
            immediate: true,
        },
    },
    methods: {
        updateUser() {
            if (!this.formDisabled) {
                this.$store.dispatch('user/update', {
                    nickname: this.newNickname,
                    pictureUrl: this.newPicture,
                })
            }
        },
    },
};
</script>
```

One can say that this solution works and he would be right. However, I see several drawbacks:

- We need to make a component responsible for the initial data load;
- A complete and complex library is needed for what seems to be a simple task.

Will the results be any better with the Composition API? Let us see!

## The same result with the Composition API?

Refactoring this application to use the Composition API should not take too long.

First, we will create the ES module that will replace our store. How can we make a module that share the data between every consumers? We can use the **singleton** design pattern:

```js
import { ref, computed } from "vue";

import { loadUserInfo, saveNewUserInfo } from './api';

const AVAILABLE_STATUS = {
    LOADING: 'LOADING',
    UPDATING: 'UPDATING',
    ERROR: 'ERROR',
};

// These data will only be created once and thus be shared by the consumers
const nickname = ref('');
const pictureUrl = ref('');
const status = ref('');

// Computed properties based on the status
const isLoading = computed(() => status.value === AVAILABLE_STATUS.LOADING);
const isUpdating = computed(() => status.value === AVAILABLE_STATUS.UPDATING);
const errorOccurred = computed(() => status.value === AVAILABLE_STATUS.ERROR);

// No need for mutations anymore, we can simply create JS methods
const apiCallReturnedWithNewUserInformation = ({ nickname: loadedNickname, pictureUrl: loadedPictureUrl }) => {
    nickname.value = loadedNickname;
    pictureUrl.value = loadedPictureUrl;
    status.value = '';
}
const load = () => {
    status.value = AVAILABLE_STATUS.LOADING;
    loadUserInfo()
        .then(apiCallReturnedWithNewUserInformation)
        .catch(() => {
            status.value = AVAILABLE_STATUS.ERROR;
        });
};

const update = (newUser) => {
    status.value = AVAILABLE_STATUS.UPDATING;
    saveNewUserInfo(newUser)
        .then(apiCallReturnedWithNewUserInformation)
        .catch(() => {
            status.value = AVAILABLE_STATUS.ERROR;
        })
};

// Fetch the user info when the module will be used for the first time
load();

// Export a method that returns every needed piece of information
export const useUserManager = () => ({
    load,
    update,
    nickname,
    pictureUrl,
    status,
    isLoading,
    isUpdating,
    errorOccurred,
});
```

Next, we need to change the way we consume data in our components:

```html
<template>
    <header>
        <template v-if="isLoading">
            User information are loading
        </template>
        <template v-else-if="isUpdating">
            User information are updating
        </template>
        <template v-else-if="errorOccurred">
            Unable to manage user information
        </template>
        <template v-else>
            Welcome {{ nickname }}
            <img :src="pictureUrl" alt="User picture" class="user-picture">
        </template>
    </header>
</template>

<script>
import { useUserManager } from './user/userManager';

export default {
    name: 'app-header',
    setup() {
        const userManager = useUserManager();
        return {
            pictureUrl: userManager.pictureUrl,
            nickname: userManager.nickname,
            isLoading: userManager.isLoading,
            isUpdating: userManager.isUpdating,
            errorOccurred: userManager.errorOccurred,
        }
    },
}
</script>

<style >
.user-picture {
    height: 40px;
    width: 40px;
    border-radius: 50%;
}
</style>
```

```html
<template>
    <form @submit.prevent="updateUser">
        <label>
            Nickname
            <input type="text" v-model="newNickname">
        </label>
        <label>
            Picture url
            <input type="text" v-model="newPicture">
        </label>
        <input type="submit" value="Validate changes" :disabled="formDisabled">
        <p v-if="errorOccurred">An error has occurred while managing user information...</p>
    </form>
</template>

<script>
import { ref, computed, watchEffect } from 'vue';
import { useUserManager } from './userManager';
export default {
    name: 'user-settings',
    setup() {
        const newNickname = ref('');
        const newPicture = ref('');

        const userManager = useUserManager();
        const formDisabled = computed(() => {
            return userManager.isLoading.value || userManager.isUpdating.value;
        });

        watchEffect(() => newNickname.value = userManager.nickname.value);
        watchEffect(() => newPicture.value = userManager.pictureUrl.value);

        const updateUser = () => {
            if (!formDisabled.value) {
                userManager.update({
                    nickname: newNickname.value,
                    pictureUrl: newPicture.value,
                });
            }
        }

        return {
            newNickname,
            newPicture,
            pictureUrl: userManager.pictureUrl,
            nickname: userManager.nickname,
            isLoading: userManager.isLoading,
            isUpdating: userManager.isUpdating,
            errorOccurred: userManager.errorOccurred,
            formDisabled,
            updateUser,
        }
    },
};
</script>
```

Finally, you can delete the store's files and remove the dependency from your application!

We now have the same result for the end-user, but our application does not depend on a library other than Vue.

I have to say, however, that this is not a silver bullet. Your application will still be hard to debug and maintain if you put everything into one big module. The Composition API is a tool, a great one, but still nothing more. It can do much more harm than good if used non-pragmatically.

-----

What we built can be seen as - and is, actually - a simple [**State Management Pattern**](https://vuejs.org/v2/guide/state-management.html#Simple-State-Management-from-Scratch). We did leverage the clarity and modularity provided by the Composition API to create what I think is the most developer and user-friendly *State Management Pattern*.

## Something ends... 

What do you think about this State Management Pattern? Will you consider using it in your application?

It was the first time that I write so much about a single subject. I hope that you have learned as much as I did along the way :)

Please send me feedback about the series and your experiences, I will be glad to read from you all!

**Comment by ThomasFerro** [comment]
TODO: Fifth article: Vue 3's Composition API and the lessons learned

Conclusion.

Add a paragraph about my take on the debate "Always keep the compatibility with previous versions" VS "Make people rewrite their applications if they want to upgrade" => I think that the rewrite of the framework was a great opportunity to flatten the API instead of copying it 1:1. However, this would have came with the cost of rewriting part of your applications. Most of the heavy-duty could have been done by a migration tool.

Edit: Not doing it for now.
